File: src/hubert_ov_seg/scripts/camera_publisher.py
#!/usr/bin/env python
import rospy
from sensor_msgs.msg import Image
from sensor_msgs.msg import CameraInfo
from cv_bridge import CvBridge
import cv2

class CameraPublisher:
    def __init__(self):
        # Initialize the ROS node
        rospy.init_node('camera_publisher', anonymous=True)

        # Create a publisher object
        self.pub_image_raw = rospy.Publisher('/camera/image_raw', Image, queue_size=1)
        self.pub_camera_info = rospy.Publisher('/camera/camera_info', CameraInfo, queue_size=1 ,latch=True)
        
        # Create a CvBridge object
        self.bridge = CvBridge()
        
        # Open the camera
        self.camera = cv2.VideoCapture(0)  # 0 for default camera
        self.camera.set(cv2.CAP_PROP_FOURCC, cv2.VideoWriter_fourcc('M', 'J', 'P', 'G'))
        if not self.camera.isOpened():
            rospy.logerr("Could not open camera")
            rospy.signal_shutdown("Camera not opened")
        
        # Set the publishing rate
        self.rate = rospy.Rate(30)  # 30 Hz

    def create_camera_info(self):
        # Initialize the CameraInfo message
        camera_info = CameraInfo()

        # Set image dimensions
        camera_info.width = 640
        camera_info.height = 480

        # Distortion coefficients: D
        # [k1, k2, p1, p2, k3]
        camera_info.distortion_model = "plumb_bob"
        camera_info.D = [
            -0.023199777932122682,  # k1
            -0.025918697269801137,  # k2
            0.001767049052526465,   # p1
            -0.0012548114794264373, # p2
            0.0                      # k3
        ]

        # Intrinsic camera matrix: K
        # [fx,  0, cx,
        #   0, fy, cy,
        #   0,  0,  1]
        camera_info.K = [
            672.8186259811123, 0.0, 313.24906274715994,
            0.0, 672.4172018457331, 256.6239831156741,
            0.0, 0.0, 1.0
        ]

        # Rectification matrix: R
        # Identity matrix since it's a mono camera
        camera_info.R = [
            1.0, 0.0, 0.0,
            0.0, 1.0, 0.0,
            0.0, 0.0, 1.0
        ]

        # Projection matrix: P
        # [fx',  0, cx', Tx,
        #   0, fy', cy', Ty,
        #   0,   0,   1,  0]
        camera_info.P = [
            668.3340553377487, 0.0, 312.55442571781094, 0.0,
            0.0, 669.8883102878623, 257.2140328011053, 0.0,
            0.0, 0.0, 1.0, 0.0
        ]

        # Optional: Set the binning and ROI if needed
        camera_info.binning_x = 0
        camera_info.binning_y = 0
        camera_info.roi.x_offset = 0
        camera_info.roi.y_offset = 0
        camera_info.roi.height = 0
        camera_info.roi.width = 0
        camera_info.roi.do_rectify = False

        # Set the frame ID and timestamp if necessary
        camera_info.header.frame_id = "camera_frame"
        camera_info.header.stamp = rospy.Time.now()

        return camera_info

    def start_publishing(self):
        while not rospy.is_shutdown():
            # Capture frame-by-frame
            ret, frame = self.camera.read()
            
            if ret:
                # Convert OpenCV image to ROS Image message
                ros_image = self.bridge.cv2_to_imgmsg(frame, encoding="bgr8")
                
                # Publish the image
                self.pub_image_raw.publish(ros_image)
                self.pub_camera_info.publish(self.create_camera_info())
            else:
                rospy.logerr("Failed to capture image")
            
            # Sleep to maintain the desired rate
            self.rate.sleep()

    def __del__(self):
        # Release the camera when the object is deleted
        self.camera.release()

if __name__ == '__main__':
    try:
        camera_publisher = CameraPublisher()
        camera_publisher.start_publishing()
    except rospy.ROSInterruptException:
        pass

File: src/hubert_ov_seg/src/camera_publisher.cpp
#include <ros/ros.h>
#include <image_transport/image_transport.h>
#include <cv_bridge/cv_bridge.h>
#include <sensor_msgs/Image.h>
#include <sensor_msgs/CameraInfo.h>
#include <opencv2/opencv.hpp>

#define CAMERA_FRAME "camera_link"

class CameraPublisher {
public:
    CameraPublisher() : nh_("~"), rate_(30) {
        image_transport::ImageTransport it(nh_);
        pub_image = it.advertiseCamera("/hubert_camera/image_raw", 1, true);

        camera_.open(0);  // Open default camera
        camera_.set(cv::CAP_PROP_FOURCC, cv::VideoWriter::fourcc('M', 'J', 'P', 'G'));

        if (!camera_.isOpened()) {
            ROS_ERROR("Could not open camera");
            ros::shutdown();
        }
    }

    ~CameraPublisher() {
        camera_.release();
    }

    sensor_msgs::CameraInfo createCameraInfo() {
        sensor_msgs::CameraInfo camera_info;
        camera_info.width = 640;
        camera_info.height = 480;
        camera_info.distortion_model = "plumb_bob";
        camera_info.D = {-0.26356547589441004, 0.38262175472105336, -0.0028605597622201632, -0.0014097888518859031, 0.0};
        camera_info.K = {1036.662690584779, 0.0, 287.6858475723572,
                        0.0, 1032.3445046229688, 222.55446891878728,
                        0.0, 0.0, 1.0};
        camera_info.R = {1.0, 0.0, 0.0,
                        0.0, 1.0, 0.0,
                        0.0, 0.0, 1.0};
        camera_info.P = {1012.5476517938175, 0.0, 286.00028020439396, 0.0,
                        0.0, 1017.6788592402461, 221.62335510675115, 0.0,
                        0.0, 0.0, 1.0, 0.0};
        camera_info.binning_x = 0;
        camera_info.binning_y = 0;
        camera_info.roi.x_offset = 0;
        camera_info.roi.y_offset = 0;
        camera_info.roi.height = 0;
        camera_info.roi.width = 0;
        camera_info.roi.do_rectify = false;
    return camera_info;
}

    void startPublishing() {
        while (ros::ok()) {
            cv::Mat frame;
            if (camera_.read(frame)) {
                sensor_msgs::ImagePtr msg = cv_bridge::CvImage(std_msgs::Header(), "bgr8", frame).toImageMsg();
                msg->header.frame_id = CAMERA_FRAME;
                msg->header.stamp = ros::Time::now();
                sensor_msgs::CameraInfoPtr info_msg(new sensor_msgs::CameraInfo(createCameraInfo()));
                info_msg->header = msg->header;
                pub_image.publish(msg, info_msg);
            } else {
                ROS_ERROR("Failed to capture image");
            }

            rate_.sleep();
            ros::spinOnce();
        }
    }

private:
    ros::NodeHandle nh_;
    image_transport::CameraPublisher pub_image;
    cv::VideoCapture camera_;
    ros::Rate rate_;
};

int main(int argc, char** argv) {
    ros::init(argc, argv, "camera_publisher");
    CameraPublisher camera_publisher;
    camera_publisher.startPublishing();
    return 0;
}